### PR TITLE
multishard_mutation_query: use coroutine::as_future

### PR DESCRIPTION
**Pull request by bhalevy**
This series converts try/catch blocks in coroutines for multishard_mutation_query to use coroutine::as_future to get and handle errors, reducing exception handling costs (that are expected on timeouts).

It was previously sent to the mailing list.
This version (v2) is just a rebase of the v1 series,
with one patch dropped as it was already merged to master independentally.

### CODE REVIEW

File: multishard_mutation_query.cc
Comment (denesb) [inline multishard_mutation_query.cc:597]: I don't see much value in coroutinizing this method as it is not helping readability. That said, this method is not on the hot-path (runs once per page), so I won't object the one added allocation. Converting this to coroutines helps uniformity (only performance critical methods remaining continuation based).

File: multishard_mutation_query.cc
Comment (denesb) [inline multishard_mutation_query.cc:764]: Eyesore nit: can we move this block into a method and instead of:
```c++
auto f = co_await coroutines::as_future(foo());
if (f.failed()) {
    ex = f.get_exception();
} else {
    f = co_await coroutines::as_future(bar()):
    if (f.failed()) {
       ex = f.get_exception(); {
    } else {
       // [...]
    }
}
```

have:
```c++
auto f = co_await coroutines::as_future(foo());
if (f.failed()) {
    return f;
}
f = co_await coroutines::as_future(bar()):
if (f.failed()) {
   return f;
}
// [...]
```
Still not as pretty as the original code but somewhat more linear. I understand why we have to do the uglification, I'm not objecting against it, just looking for a way to also keep the code readable.

**Review by denesb — COMMENTED** [review]
LGTM, just some minor style/readability suggestions.

File: multishard_mutation_query.cc
Comment (bhalevy) [reply multishard_mutation_query.cc:597]: Why added allocation?
There was a call to `do_with` before that is replaced by the coroutine frame allocation.

File: multishard_mutation_query.cc
Comment (bhalevy) [reply multishard_mutation_query.cc:764]: The lambda with save_readers can be coroutinized for simplicity as it's unlikely to fail (due to timeout).

**Comment by bhalevy** [comment]
In v3 (13edae9a31)
- Fixed cross-shard shared_ptr access (caused by invoke_on_all in v2)
- Simplified do_query

File: multishard_mutation_query.cc
Comment (denesb) [reply multishard_mutation_query.cc]: Right, I missed that, so even better.

File: multishard_mutation_query.cc
Comment (denesb) [reply multishard_mutation_query.cc:764]: Lets keep the current version instead. It doesn't have lambdas at least :).

File: multishard_mutation_query.cc
Comment (denesb) [inline multishard_mutation_query.cc:749]: We need a comment here explaining why we have this weird code. Otherwise somebody might "clean it up" in the future.

**Review by denesb — APPROVED** [review]

**Comment by bhalevy** [comment]
In v4 (5babc609c6):
- added comments around non-trivial code

**Review by denesb — APPROVED** [review]
LGTM, will merge once CI approves.

**Pull request merged**
